base/bitarray.jl

```diff
@@ -335,7 +335,7 @@ end
 @inline try_bool_conversion(x::Real) = x == 0 || x == 1 || throw(InexactError())
 @inline unchecked_bool_convert(x::Real) = x == 1
 
-function copy_to_bitarray_chunks!(Bc::Vector{UInt64}, pos_d::Int, C::Array{<:Real}, pos_s::Int, numbits::Int)
+function copy_to_bitarray_chunks!(Bc::Vector{UInt64}, pos_d::Int, C::AbstractArray{<:Real}, pos_s::Int, numbits::Int)
```

Comment:
If I remember correctly, the main reason for the `Real` restriction was that this method relies on a particular definition of the conversion to Bools (see line 330 above), as defined in "bool.jl". For general non-Real types however converting to Bools is not defined, therefore the user would have to define their own method, and the BitArray code should use the user-defined function, otherwise the behavior would be very confusing.
We could remove the restriction (and the function above) and implement two fallback methods instead:

```julia
@inline try_bool_conversion(x) = true
@inline unchecked_bool_convert(x) = Bool(x)
```

By the way we could then also add the `Complex` case in this way:

```julia
@inline try_bool_conversion(x::Complez) = isreal(x) ? try_bool_conversion(real(x)) : throw(InexactError())
@inline unchecked_bool_convert(x) = unchecked_bool_convert(real(x))
```

I'm not sure it's worth it however, I'm pretty sure I had also checked the performance at the time of writing these functions (things may have changed of course).